### ISSUE TITLE
The new __repr__ as __str__ causes infinite recursion on incorrect transforms

### ISSUE THREAD

**Issue by patricksnape** [menpo/menpo#786]
Replicate
---------
In an IPython notebook:
```
from menpo.transform import Translation
Translation([1.0, 1.0]).compose_before([1.0, 1.0])
```

Note that this code is wrong (due to the composition with a list) but the repr thing is still a bug

**Comment by jabooth** [comment]
Sorry yeah we encountered this in TPAMI crunch and just hacked it. Did you spy the recursion? Took me a few mins!

1. Invoke `__repr__` on an Menpo object without `__str__` implementation
2. Exception - no `__str__` method! Runtime gets busy preparing info for Exception message, first off, he'll need the `__repr__`...
...


I think the solution is simply to ensure all objects have a `__str__` which have the `__repr__`:

```py
    def __str__(self):
        # We have to be sure that we implement __str__ otherwise the __repr__
        # implementation below will lead to an infinite recursion.
        return 'Copyable Menpo Object with keys:\n{}'.format(
            pformat(self.__dict__))
```

I've been using this last few weeks and it seems OK - where you used to get a crash you get a pprint of the dict of the object which is actually kind of useful. Definitely raised my eyebrows though, made me realise playing with `__repr__` does have further complications...

**Comment by patricksnape** [comment]
Fixed by #787
